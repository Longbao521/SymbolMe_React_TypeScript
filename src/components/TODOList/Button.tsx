import React, { Component, ReactChild } from 'react'

interface Props {
    onClick: () => {};
    children:  ReactChild | ReactChild[];
}


export default class Button extends Component<Props> {
    state = {}

    render() {
        return (
            <div>
                <button onClick={() => this.props.onClick()}>this.props.children</button>
            </div>
        )
    }
}
