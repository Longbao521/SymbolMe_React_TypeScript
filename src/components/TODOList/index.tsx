import React, { Component } from 'react'
import Button from './Button';

interface Props {
    list: string[];
    deleteTodo: (id: number) => {};
}

export default class TODOList extends Component<Props> {
    constructor(props: Props) {
        super(props);
        this.handleTest2 = this.handleTest2.bind(this);
    }
    handleTest = () => {
        console.log('test');
    }

    handleTest2() {
        console.log('test2');
    }

    componentDidMount() {
        console.log('mount')
    }

    render() {
        return (
            <div className="todo-list">
                {this.props.list.map((todo, index) => (<div key={index}>
                    <span className="item-text ">{todo}</span>
                    <Button onClick={() => this.props.deleteTodo(index)} >done</Button>
                </div>))}
            </div>
        );
    }
}
