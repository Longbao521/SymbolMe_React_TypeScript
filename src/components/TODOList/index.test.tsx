import TODOList from './index';
import React from 'react';
import { shallow, render, mount } from 'enzyme'
import renderer from 'react-test-renderer'
const props = {
    list: ['first', 'second'],
    deleteTodo: jest.fn(),
};

const setup = () => {
    const wrapper = shallow(<TODOList {...props} />);
    return {
        props,
        wrapper,
    };
};

const setupByRender = () => {
    const wrapper = render(<TODOList {...props} />);
    return {
        props,
        wrapper,
    };
};

const setupByMount = () => {
    const wrapper = mount(<TODOList {...props} />);
    return {
        props,
        wrapper,
    };
};

it('should has Button', () => {
    const { wrapper } = setup();
    expect(wrapper.find('Button').length).toBe(2);
  });
  
  it('should render 2 item', () => {
    const { wrapper } = setupByRender();
    expect(wrapper.find('button').length).toBe(2);
  });
  
  it('should render item equal', () => {
    const { wrapper } = setupByMount();
    wrapper.find('.item-text').forEach((node, index) => {
      expect(node.text()).toBe(wrapper.props().list[index])
    });
  });
  
//   it('click item to be done', () => {
//     const { wrapper } = setupByMount();
//     wrapper.find('Button').at(0).simulate('click');
//     expect(props.deleteTodo).toBeCalled();
//   });

  it('calls componentDidMount', () => {
    const componentDidMountSpy = jest.spyOn(TODOList.prototype, 'componentDidMount');
    const { wrapper } = setup();
    expect(componentDidMountSpy).toHaveBeenCalled();
    componentDidMountSpy.mockRestore();
  });
  it('calls component handleTest', () => { // class中使用箭头函数来定义方法
    const { wrapper } = setup();
    const spyFunction = jest.spyOn(wrapper.instance(), 'handleTest');
    wrapper.instance().handleTest();
    expect(spyFunction).toHaveBeenCalled();
    spyFunction.mockRestore();
  });
  
  it('calls component handleTest2', () => { //在constructor使用bind来定义方法
    const spyFunction = jest.spyOn(TODOList.prototype, 'handleTest2');
    const { wrapper } = setup();
    wrapper.instance().handleTest2();
    expect(spyFunction).toHaveBeenCalled();
    spyFunction.mockRestore();
  });
  it('renders correctly', () => {
    const tree = renderer
    .create(<TODOList {...props} />)
            .toJSON();
  
    expect(tree).toMatchSnapshot();
  });
