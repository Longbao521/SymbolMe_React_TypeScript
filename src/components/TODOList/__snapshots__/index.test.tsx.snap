// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="todo-list"
>
  <div>
    <span
      className="item-text "
    >
      first
    </span>
    <div>
      <button
        onClick={[Function]}
      >
        this.props.children
      </button>
    </div>
  </div>
  <div>
    <span
      className="item-text "
    >
      second
    </span>
    <div>
      <button
        onClick={[Function]}
      >
        this.props.children
      </button>
    </div>
  </div>
</div>
`;
